### ISSUE TITLE
Rule blocked: CORERULES-9644 - allow use of "within" parameter with "empty" operator 

### ISSUE THREAD

**Issue by ASL-rmarshall** [cdisc-org/cdisc-rules-engine#1012]
**Links to related JIRA Tickets**
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-9644

**Rule Information**
- **Standard**: USDM
- **Rule ID**: DDF00101
- **Rule Description**: Within a study design, if study type is Interventional then at least one intervention is expected to be referenced from a procedure.

**Describe the problem**
This rule needs to report one record for each study design where none of the study interventions defined for the study design is referenced by a procedure.  I'm trying to implement this by left-joining StudyDesign -> StudyIntervention -> Procedure and then reporting the first record for each study design where `id.Procedure` is empty for all of the study design's records.  I don't think I can do this any other way because there is no (direct) reference to study design in the Procedure dataset.  Also, none of the other check operators that use `within` (or equivalent) seem to allow checking for missing values.

**Describe the solution**
At the moment, the `empty` operator assigns a value of `True` for each record in the check dataset if there are no values in the variable specified in `name`.  The `within` parameter could be added to the `empty` operator so that:
- If `within` is not specified, the operator behaves as it currently does - it indicates if the specified variable is empty for the whole dataset.
- If `within` is specified (as a list of 1 or more variable names), the check dataset would be grouped by the specified variable(s) and the result would indicate if the check variable (specified in `name`) is empty for all records in each group.

**Proposed rule logic**
```yaml
Check:
  all:
    - name: id.Procedure
      operator: empty
      within:
        - id
```

This rule is currently saved in the Production environment and positive test data is available in the USDM/DDF00101/positive/01/data folder of the CORE unitTesting SharePoint area.

**Comment by ASL-rmarshall** [comment]
It would be better if either the operator were renamed as `is_inconsistent_across_dataset` or the `True`/`False` return values were reversed to align with the name of the operator.

However, this operator is currently in use in 5 published rules:
| Core-ID     | CDISC Rule ID  | Error Message                                                                                                             | Description                                                                                                                                                                                  | Standard Name                                          | Standard Version                              | Status    |
| ----------- | -------------- | ------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------ | --------------------------------------------- | --------- |
| CORE-000699 | FB3001         | Standards units are inconsistent within the same test (category, sub-category, specimen and method)                       | Standard units should be consistent within the same assessment (having the same --TESTCD, --CAT, --SCAT, --SPEC, --METHOD values)                                                            | SDTMIG, SENDIG, SENDIG-GENETOX, SENDIG-DART, SENDIG-AR | 3.2, 3.3, 3.4, 3.0, 3.1, 3.1.1, 1.0, 1.1, 1.2 | Published |
| CORE-000689 | CG0573, FB0925 | '--TPT and --TPTNUM do not have a one-to-one relationship per unique combination of VISITNUM and --TPTREF values          | Raise an error when VISITNUM and --TPTREF are present and --TPT and --TPTNUM does not have a one-to-one relationship per unique combination of VISITNUM and --TPTREF values                  | SDTMIG                                                 | 3.2, 3.3, 3.4                                 | Published |
| CORE-000686 | CG0572, FB0924 | '--TPT and --TPTNUM do not have a one-to-one relationship per unique value of --TPTREF                                    | Raise an error when VISITNUM is not present in the dataset and --TPTREF is present in the dataset, and --TPT and --TPTNUM do not have a one-to-one relationship per unique value of --TPTREF | SDTMIG                                                 | 3.2, 3.3, 3.4                                 | Published |
| CORE-000685 | CG0571, FB0923 | '--TPT and --TPTNUM do not have a one-to-one relationship per unique value of VISITNUM                                    | Raise an error when VISITNUM is present in the dataset and --TPTREF is not present in the dataset, and --TPT and --TPTNUM do not have a one-to-one relationship per unique value of VISITNUM | SDTMIG                                                 | 3.2, 3.3, 3.4                                 | Published |
| CORE-000612 | FB7001         | Standardized units are not consistent for a given drug or metabolite across pharmacokinetic datasets within a submission. | Standardized units should be consistent for a given drug or metabolite across pharmacokinetic datasets within a submission                                                                   | SDTMIG, SENDIG, SENDIG-GENETOX, SENDIG-DART, SENDIG-AR | 3.2, 3.3, 3.4, 3.0, 3.1, 3.1.1, 1.0, 1.1, 1.2 | Published |

**Comment by ASL-rmarshall** [comment]
@nickdedonder What do you think about this?

**Comment by ASL-rmarshall** [comment]
@nickdedonder This change has now been implemented.  The 5 affected published rules are listed above and there's also 1 draft rule: SEND153 (Core Id CDISC.SENDIG.153).

**Comment by nickdedonder** [comment]
All 6 rules have been updated in the Rule editor
